conn.go

```diff
@@ -697,18 +780,66 @@ func (c *Conn) handleIncomingPacket(ctx context.Context, buf []byte, enqueue boo
 	if h.Epoch != 0 {
 		if c.state.cipherSuite == nil || !c.state.cipherSuite.IsInitialized() {
 			if enqueue {
-				c.encryptedPackets = append(c.encryptedPackets, buf)
+				c.encryptedPackets = append(c.encryptedPackets, addrPkt{rAddr, buf})
 				c.log.Debug("handshake not finished, queuing packet")
 			}
 			return false, nil, nil
 		}
 
+		// If a connection identifier had been negotiated and encryption is
+		// enabled, the connection identifier MUST be sent.
+		if len(c.state.localConnectionID) > 0 && h.ContentType != protocol.ContentTypeConnectionID {
+			c.log.Debug("discarded packet missing connection ID after value negotiated")
+			return false, nil, nil
+		}
+
 		var err error
-		buf, err = c.state.cipherSuite.Decrypt(buf)
+		hdr := recordlayer.Header{}
+		if h.ContentType == protocol.ContentTypeConnectionID {
+			hdr.ConnectionID = make([]byte, len(c.state.localConnectionID))
+		}
+		buf, err = c.state.cipherSuite.Decrypt(hdr, buf)
 		if err != nil {
 			c.log.Debugf("%s: decrypt failed: %s", srvCliStr(c.state.isClient), err)
 			return false, nil, nil
 		}
+		// If this is a connection ID record, make it look like a normal record for
+		// further processing.
+		if h.ContentType == protocol.ContentTypeConnectionID {
+			ip := &recordlayer.InnerPlaintext{}
+			if err := ip.Unmarshal(buf[h.Size():]); err != nil { //nolint:govet
+				c.log.Debugf("unpacking inner plaintext failed: %s", err)
+				return false, nil, nil
+			}
+			unpacked := &recordlayer.Header{
+				ContentType:    ip.RealType,
+				ContentLen:     uint16(len(ip.Content)),
+				Version:        h.Version,
+				Epoch:          h.Epoch,
+				SequenceNumber: h.SequenceNumber,
+			}
+			buf, err = unpacked.Marshal()
+			if err != nil {
+				c.log.Debugf("converting CID record to inner plaintext failed: %s", err)
+				return false, nil, nil
+			}
+			buf = append(buf, ip.Content...)
+		}
+
+		// If connection ID does not match discard the packet.
+		if !bytes.Equal(c.state.localConnectionID, h.ConnectionID) {
+			c.log.Debug("unexpected connection ID")
+			return false, nil, nil
+		}
+
+		// Any valid connection ID record is a candidate for updating the remote
+		// address.
+		// https://datatracker.ietf.org/doc/html/rfc9146#peer-address-update
+		if rAddr != c.RemoteAddr() {
```

Comment:
Do we need to account for condition 2 of 3 for address update?:
```
The received datagram is "newer" (in terms of both epoch and sequence number) than the newest datagram received. Reordered datagrams that are sent prior to a change in a peer address might otherwise cause a valid address change to be reverted
```

Reply:
Checking a few things here, but will follow up in a moment :)

Reply:
Following up: from my understanding, we don't currently do any reordering based on sequence number, only on epoch. The latest sequence number is [updated in the replay detector](https://github.com/pion/transport/blob/41845b068b67baa290c06efdadc59b5c5300c8ba/replaydetector/replaydetector.go#L50) when we call `markPacketAsValid()`. The [anti-replay strategy defined in the DTLS 1.2 RFC](https://datatracker.ietf.org/doc/html/rfc6347#autoid-21) is used.

I think a reasonable solution without too much churn would be to return from the callback (i.e. `markPacketAsValid()`) a `bool` that indicates whether marking the packet as valid resulted in a latest sequence number update. We could then use that value to determine whether we need to update remote address or not.

cc @Sean-Der @daenney

Reply:
I think that's a perfectly acceptable solution. Honestly if there's a unit test backing it, that's more important to me right now!

Reply:
@edaniels thanks! PR opened here with implementation https://github.com/pion/transport/pull/256

Reply:
nice. the change from there looks good here